libp2p-swarm/src/transport.rs

```diff
@@ -240,38 +251,19 @@ impl<A, B> MuxedTransport for OrTransport<A, B>
 where
 	A: MuxedTransport,
 	B: MuxedTransport,
-	A::DialAndListen: 'static,
-	B::DialAndListen: 'static,
+	A::Incoming: 'static,		// TODO: meh :-/
+	B::Incoming: 'static,		// TODO: meh :-/
 {
-	type Incoming = EitherIncomingStream<A::Incoming, B::Incoming>;
-	type Outgoing = future::Either<
-		future::Map<A::Outgoing, fn(A::RawConn) -> Self::RawConn>,
-		future::Map<B::Outgoing, fn(B::RawConn) -> Self::RawConn>,
-	>;
-	type DialAndListen = Box<Future<Item = (Self::Incoming, Self::Outgoing), Error = IoError>>;
-
-	fn dial_and_listen(self, addr: Multiaddr) -> Result<Self::DialAndListen, (Self, Multiaddr)> {
-		let (first, addr) = match self.0.dial_and_listen(addr) {
-			Ok(connec) => {
-				return Ok(Box::new(connec.map(|(inc, out)| {
-					(
-						EitherIncomingStream::First(inc),
-						future::Either::A(out.map(EitherSocket::First as fn(_) -> _)),
-					)
-				})));
-			}
-			Err(err) => err,
-		};
+	type Incoming = Box<Future<Item = (EitherSocket<A::RawConn, B::RawConn>, Multiaddr), Error = IoError>>;
 
-		match self.1.dial_and_listen(addr) {
-			Ok(connec) => Ok(Box::new(connec.map(|(inc, out)| {
-				(
-					EitherIncomingStream::Second(inc),
-					future::Either::B(out.map(EitherSocket::Second as fn(_) -> _)),
-				)
-			}))),
-			Err((second, addr)) => Err((OrTransport(first, second), addr)),
-		}
+	#[inline]
+	fn next_incoming(self) -> Self::Incoming {
```

Comment:
Why do you forward the `addr` from `first` and `second` and then `.map` it away immediately?

Reply:
It's not the address that is mapped away but a `SelectNext` object.